merkledag/ipld.md

```diff
@@ -0,0 +1,464 @@
+
+```json
+{
+  "name": "J.C.R. Licklider",
+  "name": "Hans Moravec"
+}
+```
+
+Suppose _this_ was the _exact order_ in the _Canonical Format_ (not json, but cbor), and it hashes to `QmDDD...DDD`. We would _ALWAYS_ get:
+
+```sh
+> ipld cat --json QmDDD...DDD
+{
+  "name": "J.C.R. Licklider",
+  "name": "Hans Moravec"
+}
+> ipld cat --json QmDDD...DDD/name
+"J.C.R. Licklider"
+```
+
+
+#### Path Restrictions
+
+There are some important problems that come about with path descriptions in Unix and the web. For a discussion see [this discussion](https://github.com/ipfs/go-ipfs/issues/1710). In order to be compatible with the models and expectations of unix and the web, IPLD explicitly disallows paths with certain path components. **Note that the data itself _may_ still contain these properties (someone will do it, and there are legitimate uses for it). So it is only _Path Resolvers_ that MUST NOT resolve through those paths.** The restrictions are the same as typical unix and UTF-8 path systems:
+
+
+TODO:
+- [ ] list path resolving restrictions
+- [ ] show examples
+
+#### Integers in JSON
+
+IPLD is _directly compatible_ with JSON, to take advantage of JSON's successes, but it need not be _held back_ by JSON's mistakes. This is where we can afford to follow format idiomatic choices, though care MUST be given to ensure there is always a well-defined 1:1 mapping.
+
+On the subject of integers, there exist a variety of formats which represent integers as strings in JSON, for example, [EJSON](https://www.meteor.com/ejson). These can be used and conversion to and from other formats should happen naturally-- that is, when converting JSON to CBOR, an EJSON integer should be transformed naturally to a proper CBOR integer, instead of representing it as a map with string values.
+
+
+## Serialized Data Formats
+
+IPLD supports a variety of serialized data formats trough [multicodec](https://github.com/jbenet/multicodec). These can be used however is idiomatic to the format, for example in `CBOR`, we can use `CBOR` type tags to represent the merkle-link, and avoid writing out the full string key `mlink`. Users are encouraged to use the formats to their fullest, and to store and transmit IPLD data in whatever format makes the most sense. The only requirement **is that there MUST be a well-defined one-to-one mapping with the IPLD Canonical format.** This is so that data can be transformed from one format to another, and back, without changing its meaning nor its cryptographic hashes.
+
+### Canonical Format
+
+In order to preserve merkle-linking's power, we muste ensure that there is a single **_canonical_** serialized representation of an IPLD document. This ensures that applications arrive at the same cryptographic hashes. It should be noted --though-- that this is a system-wide parameter. Future systems might change it to evolve representations. However we estimate this would need to be done no more than once per decade.
```

Comment:
@mildred i wonder if we should be specifying the canonical encoding in the expanded data itself (or at least allowing it to be overridden). storing the data will include the multicodec header, but when we parse it into logical representations (or even other wire formats), we might want to specify the canonical format.

Aside, for future thought, this is something that a `@type` could do -- similar to `@context` it could define some defaults that hold for all instances. So we _could have_ something like:

- some chosen canonical multicodec default (in our case, cbor)
- can override canonical multicodec in the object's type or context.
- can override canonical multicodec in the object itself